lib/phoenix/pubsub/redis_server.ex

```diff
@@ -0,0 +1,164 @@
+defmodule Phoenix.PubSub.RedisServer do
+  use GenServer
+  require Logger
+  alias Phoenix.PubSub.RedisServer
+
+  @moduledoc """
+  The server for the RedisAdapter
+
+  See `Phoenix.PubSub.RedisAdapter` for details and configuration options.
+  """
+
+  @derive [Access]
+  defstruct eredis_sub_pid: nil,
+            status: :disconnected,
+            reconnect_attemps: 0,
+            node_ref: nil,
+            local_pid: nil,
+            opts: []
+
+  @defaults [host: "127.0.0.1", port: 6379, password: ""]
+
+  @max_connect_attemps 3   # 15s to establish connection
+  @reconnect_after_ms 5000
+
+  @doc """
+  Starts the server
+
+  TODO document options
+  """
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts, name: Dict.fetch!(opts, :name))
+  end
+
+  @doc """
+  Initializes the server.
+
+  An initial connection establishment loop is entered. Once `:eredis_sub`
+  is started successfully, it handles reconnections automatically, so we
+  pass off reconnection handling once we find an initial connection.
+  """
+  def init(opts) do
+    {:ok, local_pid} = Phoenix.PubSub.Local.start_link()
+    opts = Dict.merge(@defaults, opts)
+    opts = Dict.merge(opts, host: String.to_char_list(to_string(opts[:host])),
+                            password: String.to_char_list(to_string(opts[:password])))
+
+    Process.flag(:trap_exit, true)
+    send(self, :establish_conn)
+
+    {:ok, struct(RedisServer, opts: opts,
+                              local_pid: local_pid,
+                              node_ref: :erlang.make_ref)}
+  end
+
+
+
+  def handle_call({:subscribe, pid, topic}, _from, state) do
+    {:reply, GenServer.call(state.local_pid, {:subscribe, pid, topic}), state}
+  end
+
+  def handle_call({:unsubscribe, pid, topic}, _from, state) do
+    {:reply, GenServer.call(state.local_pid, {:unsubscribe, pid, topic}), state}
+  end
+
+  def handle_call({:subscribers, topic}, _from, state) do
+    {:reply, GenServer.call(state.local_pid, {:subscribers, topic}), state}
+  end
+
+  def handle_call(:list, _from, state) do
+    {:reply, GenServer.call(state.local_pid, :list), state}
+  end
+
+  def handle_call({:broadcast, from_pid, topic, msg}, _from, state) do
+    with_conn state, fn state ->
+      result = :poolboy.transaction :phx_redis_pool, fn worker_pid ->
+        GenServer.call(worker_pid, {:publish_to_redis, "phx:#{topic}",
+                                   {1, state.node_ref, from_pid, msg}})
+      end
+      {:reply, result, state}
+    end
+  end
+
+  def handle_info({:pmessage, "phx:*", "phx:" <> topic, binary_msg, _client_pid}, state) do
+    :poolboy.transaction :phx_redis_pool, fn worker_pid ->
```

Comment:
So the pool here is not helping with anything because we are only using it from this server. We need to improve the design of the Redis adapter generally. Please ping me later so we can talk about it.